File: frontend/src/api/wish.ts
import { http } from "@/utils/request";

export interface Wish {
  id: number;
  user_id: number;
  author_nickname?: string;
  author_avatar?: string;
  title: string;
  content: string;
  image_urls: string[];
  category: string;
  visibility: string;
  difficulty: string;
  expected_deadline?: string | null;
  status: string;
  likes_count: number;
  completion_note?: string;
  is_anonymous: boolean;
  created_at: string;
  updated_at: string;
}

export interface ClaimSummary {
  id: number;
  wish_id: number;
  wish_title?: string;
  user_id: number;
  fulfiller_name?: string;
  progress: number;
  latest_note?: string;
  status: string;
  reject_reason?: string;
  milestone_count: number;
  created_at: string;
  updated_at: string;
}

export interface WishDetail extends Wish {
  claim?: ClaimSummary | null;
  blessing_count: number;
}

export interface CreateWishPayload {
  title: string;
  content: string;
  image_urls?: string[];
  category: string;
  visibility: string;
  difficulty: string;
  expected_deadline?: string | null;
  is_anonymous?: boolean;
}

export interface UpdateWishPayload {
  title?: string;
  content?: string;
  image_urls?: string[];
  category?: string;
  visibility?: string;
  difficulty?: string;
  expected_deadline?: string | null;
  is_anonymous?: boolean;
}

export interface PageResult<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
}

export const wishApi = {
  create: (payload: CreateWishPayload) => http.post<Wish>("/wishes", payload),
  list: (params?: Record<string, string | number | undefined>) =>
    http.get<PageResult<Wish>>("/wishes", params),
  mine: (params?: Record<string, string | number | undefined>) =>
    http.get<PageResult<Wish>>("/wishes/mine", params),
  detail: (id: number) => http.get<WishDetail>(`/wishes/${id}`),
  update: (id: number, payload: UpdateWishPayload) =>
    http.put<Wish>(`/wishes/${id}`, payload),
  remove: (id: number) => http.del<null>(`/wishes/${id}`),
  like: (id: number) => http.post<{ wish_id: number }>(`/wishes/${id}/like`),
  discover: (params?: Record<string, string | number | undefined>) =>
    http.get<PageResult<Wish>>("/discover", params),
  leaderboard: () => http.get<FulfillerStat[]>("/discover/leaderboard"),
};

export interface FulfillerStat {
  user_id: number;
  nickname: string;
  avatar: string;
  completed_count: number;
}

File: backend/internal/model/wish.go
package model

import "time"

// Wish 心愿实体。状态机：pending -> claimed -> in_progress -> completed。
type Wish struct {
	ID              uint64     `gorm:"primaryKey" json:"id"`
	UserID          uint64     `gorm:"index;not null" json:"user_id"`
	Title           string     `gorm:"size:100;not null" json:"title"`
	Content         string     `gorm:"type:text;not null" json:"content"`
	ImageURLs       string     `gorm:"type:text;not null;default:'[]'" json:"-"`
	Category        string     `gorm:"size:30;index;not null;default:other" json:"category"`
	Visibility      string     `gorm:"size:20;index;not null;default:public" json:"visibility"`
	Difficulty      string     `gorm:"size:20;not null;default:medium" json:"difficulty"`
	ExpectedDeadline *time.Time `json:"expected_deadline"`
	Status          string     `gorm:"size:20;index;not null;default:pending" json:"status"`
	LikesCount      int        `gorm:"not null;default:0" json:"likes_count"`
	CompletionNote  string     `gorm:"type:text" json:"completion_note"`
	IsAnonymous     bool       `gorm:"not null;default:false" json:"is_anonymous"`
	CreatedAt       time.Time  `json:"created_at"`
	UpdatedAt       time.Time  `json:"updated_at"`
}
